modules/tinymce/src/plugins/lists/main/ts/core/Selection.ts

```diff
@@ -67,6 +68,11 @@ const getClosestListHost = (editor: Editor, elm: Node): HTMLElement => {
   return parentBlock.getOr(editor.getBody());
 };
 
+const isListInsideLiWithSiblings = (list: SugarElement<Node>): boolean =>
```

Comment:
A couple of points

* There's code somewhere to find a parent list. I suggest using it instead of doing `Traverse.parent` and filtering.
* Once you've found a parent list, could you get the children array and check if the first / last elements are not list items? Would that give the same result? It might clean the code up a bit.

And on a wider point, thinking about the algorithm, doesn't this bug occur if there is one text element above _or_ below? It doesn't need both to break? The code requires both to activate.

Reply:
Above only would be a normal list so that would not trigger the bug:
```
<li>
  foo
  <ul>
    ..
  </ul>
</li>
```

However below ones does trigger it:
```
<li>
  <ul>
    ..
  </ul>
  foo
</li>
```

Reply:
>There's code somewhere to find a parent list. I suggest using it instead of doing Traverse.parent and filtering.
about this, the code here is intended only to check if the parent is a `li`, I refactored it after the suggestion of @spocke now it should be more clear

Reply:
Right, yes. So the code is still wrong, it needs to handle the other case.

Reply:
>  the code here is intended only to check if the parent is a li

Which is fine, but once you have the parent, if it is a `li`, then checking if the first or last children are text nodes is a very simple operation rather than doing `Traverse.prevSibling` and `Traverse.nextSibling` first.

Reply:
wait we can change the check in something like this:

```
Traverse.parent(list).exists((parent) => {
  return NodeType.isListItemNode(parent.dom)
  && (Traverse.firstChild(parent).exists((firstChild) => !NodeType.isListNode(firstChild.dom))
  || Traverse.lastChild(parent).exists((lastChild) => !NodeType.isListNode(lastChild.dom)));
});
```

and this avoid the crash for this scenario:

```
<ul>
  <li>
    <ul>
      <li>A</li>
      <li>B</li>
    </ul>
    C
  </li>
</ul>
```
but only if you try to indent the `li`s if you try to indent `C` you still have the crash

and it breaks this scenario:

```
<ul>
  <li>
    <ul>
      <li>A</li>
      <li>B</li>
    </ul>
    <p>C</p>
  </li>
</ul>
```

here doing `Indent` and then `Undent` on `B` it make it merge with `A`

the case that we want to manage here is the one with both first and last child of the `li` are not `lists` I refactored the code to fit this condition and I added a test

Reply:
That seems like an acceptable compromise. The remaining bugs can be a new ticket.